tools/check_volume_levels.m

```diff
@@ -0,0 +1,251 @@
+		sine_freqs_orig = lm.sine_freqs;
+		lm.sine_freqs = sine_freqs_orig(end:-1:1);
+		lev = level_vs_time(fn, lm);
+		pass_test = check_levels(lev, tc, lm);
+		if pass_test
+			fprintf(1,'Note: The test would pass with swapped channels.\n');
+			return
+		end
+
+		% Swapped controls?
+		lm.sine_freqs = sine_freqs_orig;
+		volumes_orig = tc.volumes;
+		tc.volumes = volumes_orig(:, end:-1:1);
+		lev = level_vs_time(fn, lm);
+		pass_test = check_levels(lev, tc, lm);
+		if pass_test
+			fprintf(1,'Note: The test would pass with swapped controls.\n')
+			return
+		end
+
+		% Swapped controls and swapped channels
+		lm.sine_freqs = sine_freqs_orig(end:-1:1);
+		lev = level_vs_time(fn, lm);
+		pass_test = check_levels(lev, tc, lm);
+		if pass_test
+			fprintf(1,'Note: The test would pass with swapped controls and swapped channels.\n')
+		end
+	end
+end
+
+function plot_levels(meas, tc, lm)
+	figure
+	plot(meas.t, meas.levels - lm.sine_dbfs);
+	grid on;
+
+	sv = size(tc.volumes);
+	hold on;
+	for j = 1:sv(2)
+		for i = 1:sv(1)
+			plot([tc.vctimes(i)+tc.meas(1) tc.vctimes(i)+tc.meas(2)], ...
+			     [tc.volumes(i,j)+tc.vtol tc.volumes(i,j)+tc.vtol], 'r--');
+			if tc.volumes(i,j) > -100
+				plot([tc.vctimes(i)+tc.meas(1) tc.vctimes(i)+tc.meas(2)], ...
+				     [tc.volumes(i,j)-tc.vtol tc.volumes(i,j)-tc.vtol], 'r--');
+			end
+		end
+	end
+	hold off;
+	xlabel('Time (s)');
+	ylabel('Gain (dB)');
+end
+
+function pass = check_levels(meas, tc, lm)
+	pass = 1;
+	gains =  meas.levels - lm.sine_dbfs;
+	sv = size(tc.volumes);
+	for j = 1:sv(2)
+		for i = 1:sv(1)
+			ts = tc.vctimes(i)+tc.meas(1);
+			te = tc.vctimes(i)+tc.meas(2);
+			idx0 = find(meas.t < te);
+			idx = find(meas.t(idx0) > ts);
+			avg_gain = mean(gains(idx, j));
+			max_gain = tc.volumes(i,j) + tc.vtol;
+			min_gain = tc.volumes(i,j) - tc.vtol;
+			if avg_gain > max_gain
+				fprintf(1, 'Channel %d Failed upper gain limit at ', j);
+				fprintf(1, '%4.1f - %4.1fs, gain %5.1f dB, max %5.1f dB\n', ...
+					ts, te, avg_gain, max_gain);
+				pass = 0;
+			end
+			if tc.volumes(i,j) > -100
+				if avg_gain < min_gain
+					fprintf(1, 'Channel %d failed lower gain limit at ', j);
+					fprintf(1, '%4.1f - %4.1fs, gain %5.1f dB, min %5.1f dB\n', ...
+						ts, te, avg_gain, min_gain);
+					pass = 0;
+				end
+			end
+		end
+	end
+end
+
+function ret = level_vs_time(fn, lm)
+	[x, fs] = audioread(fn);
+	x = bandpass_filter(x, lm.sine_freqs, fs);
+	sx = size(x);
+	tclip = sx(1) / fs;
+	nch = sx(2);
+
+	nlev = floor(tclip / lm.tgrid);
+	ngrid = lm.tgrid * fs;
+	nlength = lm.tlength * fs;
+	nmax = nlev - round(nlength / ngrid) + 1;
+	ret.t = (0:(nmax-1)) * lm.tgrid;
+	ret.levels = zeros(nmax, nch);
+	for i = 1:nmax
+		i1 = floor((i - 1) * ngrid + 1);
+		i2 = floor(i1 + nlength -1);
+		ret.levels(i, :) = level_dbfs(x(i1:i2, :));
+	end
+	ret.levels_lin = 10.^(ret.levels/20);
+end
+
+function y = bandpass_filter(x, f, fs)
+	sx = size(x);
+	y = zeros(sx(1), sx(2));
+	c1 = 0.8;
+	c2 = 1/c1;
```

Comment:
not sure if the -37dB value is related with this [0.8, 1.25] window, looks there could be some level captured when filtering 701Hz with a 1297Hz signal, 701 * [0.8, 1.25] = [560.8Hz, 876,25Hz]. So if we change this window to e.g. [0.9, 1.11] then the -37dB shift should become smaller like -50dB?

Reply:
I used quite relaxed band-pass because the filter design for IIR filters is not always robust. The narrower the band or the lower the frequencies are the more it's numerically sensitive. Octave does not support well the "z, p, k" and "sos" notation that is better than "b, a" so, to keep the function simple I kept the "b, a" filter coefficients with relaxed filter spec. The 37 dB mismatch when channels got swapped is quite large, well above error margin.

Reply:
Thanks for the info, though I am not quite understand it. :-)

Reply:
Take some time to play in Octave with butter() (lowpass, highpass,bandpass) and freqz() and you'll see the issue. For deep dive plot the  poles (complex, roots of denominator polynomial) of the transfer function and an unit circle. The narrow and very low frequency filters can have numerical coefficients precision problems that make them non-stable.

Reply:
> Take some time to play in Octave with butter() (lowpass, highpass,bandpass) and freqz() and you'll see the issue. For deep dive plot the poles (complex, roots of denominator polynomial) of the transfer function and an unit circle. The narrow and very low frequency filters can have numerical coefficients precision problems that make them non-stable.

Thanks.